Add constants for the DCA sorting modes and flags

bezin opened the pull request:
This PR adds constants for sorting mode and flags, that can be used in the data container arrays. This gives us a little bit more autocomplete in IDEs and does not need us to remember every single mode and flag or look it up in the docs every time ;-)

bezin committed:
Add constants for DCA sorting mode and flags

m-vo commented:
That's a really helpful addition!

I'd put it into `DC_Table` though, as it's only relevant there.  E.g. with a prefix similar to the config key name: `SORTING_MODE__TREE`, `SORTING_FLAG__INITIAL_LETTER_ASC` and so on.

fritzmg commented:
I prefer having separate classes, rather than packing everything into `DC_Table`.

m-vo commented:
> I prefer having separate classes, rather than packing everything into DC_Table.

But then you need to know the classes…

bezin commented:
For me typing `SortingMode` or `SortingFlag` came more naturally when I am in the  DCA `sorting` section configuring `mode` or `sorting`. Also, I did not have a Kneifzange at hand to work on `DC_Table` 😛 

I would prefer this: 
<img width="207" alt="2021-03-19 14_20_04" src="https://user-images.githubusercontent.com/4400435/111787275-0b05e980-88bf-11eb-96c4-0bf17be5f7b7.png">

over this:
<img width="215" alt="2021-03-19 14_19_47" src="https://user-images.githubusercontent.com/4400435/111787270-0a6d5300-88bf-11eb-8c76-b2856dc413d8.png">

fritzmg commented:
> But then you need to know the classes…

But that's also the case with `DC_Table`. You don't write DCAs within the `DC_Table` or `DataContainer` class, so you wouldn't be able to just type `self::` anyway. You would need to have knowledge about where the flags are stored in any case.

aschempp commented:
This way, I have to remember 1,2,3 different class names! I would prefer it in `DataContainer` as well, e.g. `DataContainer::SORT_ INITIAL_LETTER_ASC`.

rabauss commented:
I'm torn! On the one hand, you have to remember the classes, but on the other hand, what are the other large lists of constants for DCAs? Actually only the input types, right? If we add it to the `DataContainer` it should be added via traits!

bezin commented:
Yes, good point. I would like to add these as well. Wouldn't it make sense, to add the input type constants to every widget class? Which, I must admit, would make the case for adding the sorting mode and flag constants to the DataContainer/DC_Table 🤔

leofeyer commented:
> This way, I have to remember 1,2,3 different class names! I would prefer it in `DataContainer`

Me too.

aschempp commented:
> Wouldn't it make sense, to add the input type constants to every widget class? 

It would actually make sense to use the class name instead of an arbitrary string for the widget types… not sure if we can add that in a BC-way 🤔

fritzmg commented:
> It would actually make sense to use the class name instead of an arbitrary string for the widget types…

I thought we did not want to support this: https://github.com/contao/contao/pull/2378#issuecomment-714511735

bezin commented:
I deleted the new classes and moved everything to `DC_Table` – as @m-vo already mentioned, the constants are only used there but not in `DC_Folder` or `DC_File`.

leofeyer commented:
> I'd put it into `DC_Table` though, as it's only relevant there. E.g. with a prefix similar to the config key name: `SORTING_MODE__TREE`, `SORTING_FLAG__INITIAL_LETTER_ASC` and so on.

I don't like this. Yes, the constants might not be used in DC_Folder, but they might be used in a custom child class of DC_Folder. In the end, it is more about in the scope in which the constants make sense – and that is definitely in any data container. Thus `DataContainer::SORT_ASC` instead of `DC_Table::SORT_ASC`. (Files could be sorted descending, too.)

leofeyer commented:
As discussed in Mumble on April 8th, two things should be changed:

1. The constants should be in the `DataContainer` class.
2. The constants should be used in the code. 😄

bezin commented:
Hello there, thanks for your comments (unfortunately I could not make it to the Mumble call as I was on my holidays). 
I updated as you discussed. I hope I got them all – please, somebody double check :-)

One minor thing came up: The sorting flags `3` and `4` differ a little between the list sorting flag and the per-field sorting flag: For the list it is always `Sort by initial first two letters` (hence the name), but for a single field you may configure how many letters should be taken into account with the `length` option.

I still used the sorting flag there, as I thought everything else would be more confusing. If somebody has an idea for a name that reflects both possibilities, we could update the naming.

bezin commented:
Hi there, I actually thought that this was finished already. 

Anyways, I renamed `SORT_INITIAL_TWO_LETTERS_ASC` to `SORT_INITIAL_LETTERS_ASC` so we now have the following options:

* **SORT_INITIAL_LETTER_ASC (or_DESC)**
  Sort by the initial letter
* **SORT_INITIAL_LETTERS_ASC (or _DESC)**
   Sort by initial letters, which is either fixed to the first two characters in list sorting or can be configured with `length` when used in a field configuration.

It's final from my side now, let me know if there is anything I can do.

Cheers 👋

aschempp commented on core-bundle/src/Resources/contao/classes/DataContainer.php:
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
+	/**
+	 * Records are sorted by a fixed field
+	 */
+	public const MODE_FIXED_FIELD = 1;
```
```suggestion
	public const MODE_SORTED = 1;
```

aschempp commented on core-bundle/src/Resources/contao/classes/DataContainer.php:
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
+	/**
+	 * Records are sorted by a switchable field
+	 */
+	public const MODE_SWITCHABLE_FIELD = 2;
```
```suggestion
	public const MODE_SORTABLE = 2;
```

aschempp commented on core-bundle/src/Resources/contao/classes/DataContainer.php:
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
+	/**
+	 * Records are sorted by the parent table
+	 */
+	public const MODE_PARENT_TABLE = 3;
```
```suggestion
	public const MODE_SORT_PARENT = 3;
```

aschempp commented on core-bundle/src/Resources/contao/classes/DataContainer.php:
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
+	/**
+	 * Displays the child records of a parent record (see style sheets module)
+	 */
+	public const MODE_CHILD_RECORD = 4;
```
```suggestion
	public const MODE_CHILDREN = 4;
```

aschempp commented on core-bundle/src/Resources/contao/classes/DataContainer.php:
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
+	public const MODE_PARENT_TABLE = 3;
+
+	/**
+	 * Displays the child records of a parent record (see style sheets module)
```
```suggestion
	 * Displays the child records of a parent record (see content elements)
```

bezin commented:
I do like all of your suggestions 👍 Will update them later.

bezin commented:
Everything is updated.

leofeyer reviewed (changes requested):
I am not happy with `MODE_CHILDREN` and `MODE_CHILD_TREE`. Can we come up with something else? @contao/developers /cc

leofeyer commented:
@bezin Can you please rebase the PR on the current `4.x` branch and resolve the conflicts?

bezin commented:
What about `MODE_CHILD_TREE` and `MODE_CHILD_LIST` instead of `MODE_CHILDREN`?

aschempp reviewed (changes requested):
```php
public const MODE_UNSORTED = 0;
public const MODE_SORTED = 1;
public const MODE_SORTABLE = 2;
public const MODE_SORTED_PARENT = 3;
public const MODE_PARENT = 4;
public const MODE_TREE = 5;
public const MODE_TREE_EXTENDED = 6;
```

bezin commented:
All done here from my point of view 👍

Toflar reviewed (dismissed):
Nice work @bezin <3

Toflar commented:
Need to fix the CI tests though ;)

bytehead commented on core-bundle/src/Resources/contao/drivers/DC_Table.php:
```diff
@@ -4800,7 +4800,7 @@ protected function listView()
 <a href="' . $this->getReferer(true, $this->ptable) . '" class="header_back" title="' . StringUtil::specialchars($GLOBALS['TL_LANG']['MSC']['backBTTitle']) . '" accesskey="b" onclick="Backend.getScrollOffset()">' . $GLOBALS['TL_LANG']['MSC']['backBT'] . '</a> ' : (isset($GLOBALS['TL_DCA'][$this->strTable]['config']['backlink']) ? '
 <a href="contao/main.php?' . $GLOBALS['TL_DCA'][$this->strTable]['config']['backlink'] . '" class="header_back" title="' . StringUtil::specialchars($GLOBALS['TL_LANG']['MSC']['backBTTitle']) . '" accesskey="b" onclick="Backend.getScrollOffset()">' . $GLOBALS['TL_LANG']['MSC']['backBT'] . '</a> ' : '')) . ((Input::get('act') != 'select' && !($GLOBALS['TL_DCA'][$this->strTable]['config']['closed']  ...
-<a href="' . ($this->ptable ? $this->addToUrl('act=create' . ((($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) < 4) ? '&amp;mode=2' : '') . '&amp;pid=' . $this->intId) : $this->addToUrl('act=create')) . '" class="header_new" title="' . StringUtil::specialchars($labelNew[1] ?? '') . '" accesskey="n" onclick="Backend.getScrollOffset()">' . $labelNew[0] . '</a> ' : '') . $th ...
+<a href="' . ($this->ptable ? $this->addToUrl('act=create' . ((($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) < self::MODE_PARENT ? '&amp;mode=2' : '') . '&amp;pid=' . $this->intId) : $this->addToUrl('act=create')) . '" class="header_new" title="' . StringUtil::specialchars($labelNew[1] ?? '') . '" accesskey="n" onclick="Backend.getScrollOffset()">' . $labelNew[0] . '</a ...
```
There's a `)` missing after `self::MODE_PARENT`

bytehead replied:
```suggestion
<a href="' . ($this->ptable ? $this->addToUrl('act=create' . ((($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) < self::MODE_PARENT) ? '&amp;mode=2' : '') . '&amp;pid=' . $this->intId) : $this->addToUrl('act=create')) . '" class="header_new" title="' . StringUtil::specialchars($labelNew[1] ?? '') . '" accesskey="n" onclick="Backend.getScrollOffset()">' . $labelNew[0] . '</a> ' : '') . $this->generateGlobalButtons() . '
```

bezin replied:
Thanks, just spotted this myself 👍

bezin commented:
Done 👍

leofeyer commented:
I would also change this: https://github.com/bezin/contao/pull/1

leofeyer committed:
Put the constants before the class constants

leofeyer commented:
Thank you @bezin.
